packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1991,8 +2598,241 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
+          nextPage.createRoute(context),
+          initialState: _RouteLifecycle.staging,
+        );
+        assert(
+          newEntry.route.settings == nextPage,
+          'If a route is created from a page, its must have that page as its '
+          'settings.',
+        );
+        newHistory.add(newEntry);
+      } else {
+        // Removes the key from pageKeyToOldEntry to indicate it is taken.
+        final _RouteEntry matchingEntry = pageKeyToOldEntry.remove(nextPage.key);
+        assert(matchingEntry.canUpdateFrom(nextPage));
+        matchingEntry.route._updateSettings(nextPage);
+        newHistory.add(matchingEntry);
+      }
+    }
+
+    // Any remaining old routes that do not have a match will need to be removed.
+    final Map<StageableRoute, StageableRoute> locationToExitingPageRoute = <StageableRoute, StageableRoute>{};
+    while (oldEntriesBottom <= oldEntriesTop) {
+      final _RouteEntry potentialEntryToRemove = _history[oldEntriesBottom];
+      oldEntriesBottom += 1;
+
+      if (!potentialEntryToRemove.hasPage) {
+        assert(previousOldPageRouteEntry != null);
+        final List<_RouteEntry> pagelessRoutes = pageRouteToPagelessRoutes
+          .putIfAbsent(
+          previousOldPageRouteEntry,
+            () => <_RouteEntry>[]
+        );
+        pagelessRoutes.add(potentialEntryToRemove);
+        assert(() {
+          potentialEntryToRemove._debugIsExiting = previousOldPageRouteEntry._debugIsExiting;
+          return true;
+        }());
+        continue;
+      }
+
+      final Page<dynamic> potentialPageToRemove = potentialEntryToRemove.route.settings as Page<dynamic>;
+      // Marks for transition delegate to remove if this old page does not have
+      // a key or was not taken during updating the middle of new page.
+      if (
+        potentialPageToRemove.key == null ||
+        pageKeyToOldEntry.containsKey(potentialPageToRemove.key)
+      ) {
+        locationToExitingPageRoute[previousOldPageRouteEntry] = potentialEntryToRemove;
+        assert(() {
+          potentialEntryToRemove._debugIsExiting = true;
+          return true;
+        }());
+      }
+      previousOldPageRouteEntry = potentialEntryToRemove;
+    }
+
+    // We've scanned the whole list.
+    assert(oldEntriesBottom == oldEntriesTop + 1);
+    assert(newPagesBottom == newPagesTop + 1);
+    newPagesTop = widget.pages.length - 1;
+    oldEntriesTop = _history.length - 1;
+    // Verifies we either reach the bottom or the oldEntriesBottom must be updatable
+    // by newPagesBottom.
+    assert(() {
+      if (oldEntriesBottom <= oldEntriesTop)
+        return newPagesBottom <= newPagesTop &&
+          _history[oldEntriesBottom].hasPage &&
+          _history[oldEntriesBottom].canUpdateFrom(widget.pages[newPagesBottom]);
+      else
+        return newPagesBottom > newPagesTop;
+    }());
+
+    // Updates the top of the list.
+    while ((oldEntriesBottom <= oldEntriesTop) && (newPagesBottom <= newPagesTop)) {
```

Comment:
Why can we not do this directly in the second while loop of this function?

Reply:
if we do that we will have to store it into a list in reverse order and them combine at the end. It need to iterate through them twice anyway. This method will save more space, and this is actually following the same algorithm as updateChildren